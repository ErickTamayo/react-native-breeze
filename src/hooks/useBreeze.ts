import { BreezeStyle } from "../plugin/helpers/styles";
import { useMediaStyle } from "./useMediaStyle";
import { useBreezeStyle } from "./useBreezeStyle";

export const useBreeze = (breezeStyle: BreezeStyle) => {
  const mediaStyle = useBreezeStyle(breezeStyle);
  const variantStyle = useMediaStyle(mediaStyle);
  // const { hoverStyle, onMouseEnter, onMouseLeave } = useHover(variantStyle);
  // const { focusStyle, onFocus, onBlur } = useFocus(variantStyle);
  // const style = useCombineStyles(variantStyle, hoverStyle, focusStyle);

  // return { style: variantStyle, onMouseEnter, onMouseLeave, onFocus, onBlur };
  return { style: variantStyle };
};
