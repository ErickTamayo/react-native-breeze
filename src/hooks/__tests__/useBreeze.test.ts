import { renderHook, act } from "@testing-library/react-hooks";
import { Platform, Dimensions } from "react-native";
import useBreeze from "../useBreeze";

describe("useBreeze", () => {
  afterAll(() => {
    Platform.OS = "ios";
  });

  it("should parse the correct regular style", () => {
    const { result } = renderHook(() => useBreeze());
    expect(result.current.parse("bg-gray-500")).toMatchSnapshot();
  });

  describe.each([
    undefined,
    "native:",
    "ios:",
    "android:",
    "macos:",
    "windows:",
    "web:",
  ])("platform %s", (platform = "") => {
    beforeEach(() => {
      Platform.OS = platform.replace(":", "") as any;
    });

    it("should parse the correct style", () => {
      const { result } = renderHook(() => useBreeze());
      const style = `${platform}w-0`;
      expect(result.current.parse(style)).toEqual({ width: 0 });
    });

    describe.each([
      [undefined, 0],
      ["sm:", 640],
      ["md:", 768],
      ["lg:", 1024],
      ["xl:", 1280],
    ])("media %s", (media = "", width) => {
      beforeEach(() => {
        (Dimensions as any).get = jest.fn(() => ({ height: 1, width }));
      });

      it("should parse the correct media style", () => {
        const { result } = renderHook(() => useBreeze());
        const style = `${platform}${media}:w-0`;
        expect(result.current.parse(style)).toEqual({ width: 0 });
      });

      it(`should parse the correct hovered style`, () => {
        const { result } = renderHook(() => useBreeze());
        const style = `${platform}${media}hover:w-0`;
        const onMouseEnter = jest.fn();
        const onMouseLeave = jest.fn();

        act(() => {
          result.current.handleOnMouseEnter(onMouseEnter)({});
        });

        expect(result.current.parse(style)).toEqual({ width: 0 });
        expect(onMouseEnter).toHaveBeenCalled();

        act(() => {
          result.current.handleOnMouseLeave(onMouseLeave)({});
        });

        expect(result.current.parse(style)).toEqual({});
        expect(onMouseLeave).toHaveBeenCalled();
      });

      it(`should parse the correct focused style`, () => {
        const { result } = renderHook(() => useBreeze());
        const style = `${platform}${media}focus:w-0`;
        const onFocus = jest.fn();
        const onBlur = jest.fn();

        act(() => {
          result.current.handleOnFocus(onFocus)({});
        });

        expect(result.current.parse(style)).toEqual({ width: 0 });
        expect(onFocus).toHaveBeenCalled();

        act(() => {
          result.current.handleOnBlur(onBlur)({});
        });

        expect(result.current.parse(style)).toEqual({});
        expect(onBlur).toHaveBeenCalled();
      });

      if (["ios:", "android:"].includes(platform)) {
        describe.each([
          ["landscape", (width || 1) * 2, width || 1],
          ["portrait", width || 1, (width || 1) * 2],
        ])("%s", (orientation, width, height) => {
          beforeEach(() => {
            (Dimensions as any).get = jest.fn(() => ({ height, width }));
          });

          it(`should parse the correct style in ${orientation}`, () => {
            const { result } = renderHook(() => useBreeze());
            const style = `${platform}${media}${orientation}:w-0`;

            expect(result.current.parse(style)).toEqual({ width: 0 });
          });
        });
      }
    });
  });
});
