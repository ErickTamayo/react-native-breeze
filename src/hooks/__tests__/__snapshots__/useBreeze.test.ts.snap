// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useBreeze should parse the correct regular style 1`] = `
Object {
  "backgroundColor": "#a0aec0",
}
`;
